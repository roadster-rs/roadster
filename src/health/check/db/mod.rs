#[cfg(feature = "db-diesel-pool")]
pub mod diesel;
#[cfg(feature = "db-diesel-mysql-pool-async")]
pub mod diesel_mysql_async;
#[cfg(feature = "db-diesel-postgres-pool-async")]
pub mod diesel_pg_async;
#[cfg(feature = "db-sea-orm")]
pub mod sea_orm;
