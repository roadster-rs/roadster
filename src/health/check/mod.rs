use std::time::Duration;

pub mod db;
pub mod default;
#[cfg(feature = "email")]
pub mod email;
pub mod registry;
#[cfg(feature = "worker-sidekiq")]
pub mod sidekiq_enqueue;
#[cfg(feature = "worker-sidekiq")]
pub mod sidekiq_fetch;

use crate::error::RoadsterResult;
use async_trait::async_trait;
#[cfg(feature = "open-api")]
use schemars::JsonSchema;
use serde_derive::{Deserialize, Serialize};
use serde_json::Value;
use serde_with::{serde_as, skip_serializing_none};
use tracing::error;
use typed_builder::TypedBuilder;

#[serde_as]
#[skip_serializing_none]
#[derive(Debug, Clone, Serialize, Deserialize, TypedBuilder)]
#[cfg_attr(feature = "open-api", derive(JsonSchema))]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CheckResponse {
    pub status: Status,
    /// Total latency of checking the health of the resource in milliseconds.
    #[builder(setter(transform = |duration: std::time::Duration| duration.as_millis()))]
    pub latency: u128,
    /// Custom health data, for example, separate latency measurements for acquiring a connection
    /// from a resource pool vs making a request with the connection.
    #[builder(
        default,
        setter(transform = |custom: impl serde::Serialize| serialize_custom(custom))
    )]
    pub custom: Option<Value>,
}

fn serialize_custom(custom: impl serde::Serialize) -> Option<Value> {
    Some(
        serde_json::to_value(custom)
            .unwrap_or_else(|err| Value::String(format!("Unable to serialize custom data: {err}"))),
    )
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "open-api", derive(JsonSchema))]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub enum Status {
    Ok,
    Err(ErrorData),
}

#[derive(Debug, Clone, Serialize, Deserialize, TypedBuilder)]
#[cfg_attr(feature = "open-api", derive(JsonSchema))]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct ErrorData {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default, setter(strip_option))]
    pub msg: Option<String>,
}

/// Trait used to check the health of the app before its services start up.
///
/// This is a separate trait, vs adding a "health check" method to [`crate::service::AppService`],
/// to allow defining health checks that apply to multiple services. For example, most services
/// would require the DB and Redis connections to be valid, so we would want to perform a check for
/// these resources a single time before starting any service instead of once for every service that
/// needs the resources.
///
/// Another benefit of using a separate trait is, because the health checks are decoupled from
/// services, they can potentially be used in other parts of the app. For example, they can
/// be used to implement the "health check" API endpoint.
#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub trait HealthCheck: Send + Sync {
    /// The name of the [`HealthCheck`].
    fn name(&self) -> String;

    /// Whether the health check is enabled. If the health check is not enabled, Roadster will not
    /// run it. However, if a consumer wants, they can certainly create a [`HealthCheck`] instance
    /// and directly call [`HealthCheck::check`] even if [`HealthCheck::enabled`] returns `false`.
    fn enabled(&self) -> bool;

    /// Run the [`HealthCheck`].
    // Note: This is not able to take a state/AppContext type parameter because that makes it
    // not "dyn-compatible", which means it can't be made into an object. If a `HealthCheck` impl
    // needs the state/AppContext, it needs to have it as a field in its struct, and it should
    // use an `AppContextWeak` to avoid a reference cycle.
    async fn check(&self) -> RoadsterResult<CheckResponse>;
}

// This method is not used in all feature configurations.
#[allow(dead_code)]
fn missing_context_response() -> CheckResponse {
    let msg = "AppContext missing; is the app shutting down?".to_string();
    error!(msg);
    CheckResponse::builder()
        .status(Status::Err(ErrorData::builder().msg(msg).build()))
        .latency(Duration::from_secs(0))
        .build()
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;

    #[test]
    fn missing_context_response() {
        assert_json_snapshot!(super::missing_context_response());
    }
}
