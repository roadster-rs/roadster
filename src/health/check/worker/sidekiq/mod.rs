pub mod sidekiq_enqueue;
pub mod sidekiq_fetch;
