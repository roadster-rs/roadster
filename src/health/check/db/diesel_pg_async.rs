use crate::api::core::health::db_diesel_health_pg_async;
use crate::app::context::{AppContext, AppContextWeak};
use crate::error::RoadsterResult;
use crate::health::check::{missing_context_response, CheckResponse, HealthCheck};
use async_trait::async_trait;
use tracing::instrument;

pub struct DbDieselPgAsyncHealthCheck {
    pub(crate) context: AppContextWeak,
}

#[async_trait]
impl HealthCheck for DbDieselPgAsyncHealthCheck {
    fn name(&self) -> String {
        "db-diesel-postgres-async".to_string()
    }

    fn enabled(&self) -> bool {
        self.context
            .upgrade()
            .map(|context| enabled(&context))
            .unwrap_or_default()
    }

    #[instrument(skip_all)]
    async fn check(&self) -> RoadsterResult<CheckResponse> {
        let context = self.context.upgrade();
        let response = match context {
            Some(context) => db_diesel_health_pg_async(&context, None).await,
            None => missing_context_response(),
        };
        Ok(response)
    }
}

fn enabled(context: &AppContext) -> bool {
    context
        .config()
        .health_check
        .database
        .common
        .enabled(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;
        config.health_check.database.common.enable = enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        // Act/Assert
        assert_eq!(super::enabled(&context), expected_enabled);
    }
}
