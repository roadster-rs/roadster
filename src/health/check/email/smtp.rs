use crate::api::core::health::smtp_health;
use crate::app::context::{AppContext, AppContextWeak};
use crate::error::RoadsterResult;
use crate::health::check::{missing_context_response, CheckResponse, HealthCheck};
use async_trait::async_trait;
use tracing::instrument;

pub struct SmtpHealthCheck {
    pub(crate) context: AppContextWeak,
}

#[async_trait]
impl HealthCheck for SmtpHealthCheck {
    fn name(&self) -> String {
        "smtp".to_string()
    }

    fn enabled(&self) -> bool {
        self.context
            .upgrade()
            .map(|context| enabled(&context))
            .unwrap_or_default()
    }

    #[instrument(skip_all)]
    async fn check(&self) -> RoadsterResult<CheckResponse> {
        let context = self.context.upgrade();
        let response = match context {
            Some(context) => smtp_health(&context, None).await,
            None => missing_context_response(),
        };
        Ok(response)
    }
}

fn enabled(context: &AppContext) -> bool {
    context.config().health_check.smtp.common.enabled(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;
        config.health_check.smtp.common.enable = enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        // Act/Assert
        assert_eq!(super::enabled(&context), expected_enabled);
    }
}
