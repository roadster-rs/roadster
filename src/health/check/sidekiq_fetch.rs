use crate::api::core::health::redis_health;
use crate::app::context::{AppContext, AppContextWeak};
use crate::error::RoadsterResult;
use crate::health::check::{CheckResponse, HealthCheck, missing_context_response};
use async_trait::async_trait;
use tracing::instrument;

pub struct SidekiqFetchHealthCheck {
    pub(crate) context: AppContextWeak,
}

#[async_trait]
impl HealthCheck for SidekiqFetchHealthCheck {
    fn name(&self) -> String {
        "sidekiq-fetch".to_string()
    }

    fn enabled(&self) -> bool {
        self.context
            .upgrade()
            .map(|context| enabled(&context))
            .unwrap_or_default()
    }

    #[instrument(skip_all)]
    async fn check(&self) -> RoadsterResult<CheckResponse> {
        let context = self.context.upgrade();
        let response = match context {
            Some(context) => {
                let redis = context.redis_fetch().as_ref().ok_or_else(|| {
                    crate::error::sidekiq::SidekiqError::Message(
                        "Redis fetch connection pool is not present".to_owned(),
                    )
                })?;
                redis_health(redis, None).await
            }
            None => missing_context_response(),
        };
        Ok(response)
    }
}

fn enabled(context: &AppContext) -> bool {
    context.redis_fetch().is_some()
        && context
            .config()
            .health_check
            .sidekiq
            .common
            .enabled(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use bb8::Pool;
    use rstest::rstest;
    use sidekiq::RedisConnectionManager;

    #[rstest]
    #[case(false, Some(true), true, true)]
    #[case(false, Some(true), false, false)]
    #[case(false, Some(false), false, false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    #[tokio::test]
    async fn enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] pool: bool,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;
        config.health_check.sidekiq.common.enable = enable;

        let redis_fetch_pool = if pool {
            let redis_fetch = RedisConnectionManager::new("redis://invalid_host:1234").unwrap();
            let pool = Pool::builder().build_unchecked(redis_fetch);
            Some(pool)
        } else {
            None
        };
        let context = AppContext::test(Some(config), None, redis_fetch_pool).unwrap();

        // Act/Assert
        assert_eq!(super::enabled(&context), expected_enabled);
    }
}
