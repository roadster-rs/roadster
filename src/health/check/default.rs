use crate::app::context::AppContext;
use crate::health::check::HealthCheck;
#[cfg(feature = "db-diesel-mysql-pool-async")]
use crate::health::check::db::diesel_mysql_async::DbDieselMysqlAsyncHealthCheck;
#[cfg(feature = "db-diesel-postgres-pool-async")]
use crate::health::check::db::diesel_pg_async::DbDieselPgAsyncHealthCheck;
#[cfg(feature = "db-sea-orm")]
use crate::health::check::db::sea_orm::DbSeaOrmHealthCheck;
#[cfg(feature = "email-smtp")]
use crate::health::check::email::smtp::SmtpHealthCheck;
#[cfg(feature = "worker-pg")]
use crate::health::check::worker::pg::PgWorkerHealthCheck;
#[cfg(feature = "worker-sidekiq")]
use crate::health::check::worker::sidekiq::sidekiq_enqueue::SidekiqEnqueueHealthCheck;
#[cfg(feature = "worker-sidekiq")]
use crate::health::check::worker::sidekiq::sidekiq_fetch::SidekiqFetchHealthCheck;
use std::collections::BTreeMap;
use std::sync::Arc;

pub fn default_health_checks(
    #[allow(unused_variables)] context: &AppContext,
) -> BTreeMap<String, Arc<dyn HealthCheck>> {
    let health_checks: Vec<Arc<dyn HealthCheck>> = vec![
        #[cfg(feature = "db-sea-orm")]
        Arc::new(DbSeaOrmHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "db-diesel-postgres-pool")]
        Arc::new(crate::health::check::db::diesel::DbDieselHealthCheck::new(
            context,
            "postgres",
            |context| context.diesel_pg_pool(),
        )),
        #[cfg(feature = "db-diesel-mysql-pool")]
        Arc::new(crate::health::check::db::diesel::DbDieselHealthCheck::new(
            context,
            "mysql",
            |context| context.diesel_mysql_pool(),
        )),
        #[cfg(feature = "db-diesel-sqlite-pool")]
        Arc::new(crate::health::check::db::diesel::DbDieselHealthCheck::new(
            context,
            "sqlite",
            |context| context.diesel_sqlite_pool(),
        )),
        #[cfg(feature = "db-diesel-postgres-pool-async")]
        Arc::new(DbDieselPgAsyncHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "db-diesel-mysql-pool-async")]
        Arc::new(DbDieselMysqlAsyncHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "worker-sidekiq")]
        Arc::new(SidekiqEnqueueHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "worker-sidekiq")]
        Arc::new(SidekiqFetchHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "worker-pg")]
        Arc::new(PgWorkerHealthCheck {
            context: context.downgrade(),
        }),
        #[cfg(feature = "email-smtp")]
        Arc::new(SmtpHealthCheck {
            context: context.downgrade(),
        }),
    ];

    health_checks
        .into_iter()
        .filter(|check| check.enabled())
        .map(|check| (check.name(), check))
        .collect()
}

#[cfg(all(
    test,
    feature = "worker-sidekiq",
    feature = "worker-pg",
    feature = "db-sea-orm",
    feature = "db-diesel-postgres-pool",
    feature = "db-diesel-mysql-pool",
    feature = "db-diesel-sqlite-pool",
    feature = "db-diesel-postgres-pool-async",
    feature = "db-diesel-mysql-pool-async",
    feature = "email-smtp"
))]
mod tests {
    use crate::app::context::AppContext;
    use crate::config::AppConfig;
    use crate::testing::snapshot::TestCase;
    use bb8::Pool;
    use insta::assert_toml_snapshot;
    use itertools::Itertools;
    use rstest::{fixture, rstest};
    use sidekiq::RedisConnectionManager;

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(false)]
    #[case(true)]
    #[tokio::test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    async fn default_health_checks(_case: TestCase, #[case] default_enable: bool) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;

        let redis_fetch = RedisConnectionManager::new("redis://invalid_host:1234").unwrap();
        let pool = Pool::builder().build_unchecked(redis_fetch);

        let context = AppContext::test(Some(config), None, Some(pool)).unwrap();

        // Act
        let health_checks = super::default_health_checks(&context);
        let health_checks = health_checks.keys().collect_vec();

        // Assert
        assert_toml_snapshot!(health_checks);
    }
}
