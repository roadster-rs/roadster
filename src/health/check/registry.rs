use crate::app::context::AppContext;
use crate::error::RoadsterResult;
use crate::health::check::HealthCheck;
use crate::health::check::default::default_health_checks;
use std::collections::BTreeMap;
use std::sync::Arc;
use thiserror::Error;
use tracing::info;

#[derive(Debug, Error)]
#[non_exhaustive]
pub enum HealthCheckRegistryError {
    /// The provided [`HealthCheck`] was already registered. Contains the [`HealthCheck::name`]
    /// of the provided service.
    #[error("The provided `HealthCheck` was already registered: `{0}`")]
    AlreadyRegistered(String),

    #[error(transparent)]
    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// Registry for [`HealthCheck`]s that will be run in the app.
///
/// Health checks are used in multiple parts of the app, for example:
/// 1. As pre-boot checks to ensure the app's resource dependencies are healthy.
/// 2. As a "core" API that can be used from multiple components, e.g. the `_health` HTTP endpoint
///    and the health CLI command.
pub struct HealthCheckRegistry {
    health_checks: BTreeMap<String, Arc<dyn HealthCheck>>,
}

impl HealthCheckRegistry {
    pub(crate) fn new(context: &AppContext) -> Self {
        Self {
            health_checks: default_health_checks(context),
        }
    }

    pub fn register<H>(&mut self, health_check: H) -> RoadsterResult<()>
    where
        H: HealthCheck + 'static,
    {
        self.register_arc(Arc::new(health_check))
    }

    pub fn register_arc(&mut self, health_check: Arc<dyn HealthCheck>) -> RoadsterResult<()> {
        let name = health_check.name();

        if !health_check.enabled() {
            info!(health_check.name=%name, "Health check is not enabled, skipping registration");
            return Ok(());
        }

        info!(health_check.name=%name, "Registering health check");

        if self
            .health_checks
            .insert(name.clone(), health_check)
            .is_some()
        {
            return Err(HealthCheckRegistryError::AlreadyRegistered(name).into());
        }
        Ok(())
    }

    pub fn checks(&self) -> Vec<Arc<dyn HealthCheck>> {
        self.health_checks.values().cloned().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use crate::health::check::MockHealthCheck;
    use rstest::rstest;

    #[rstest]
    #[case(true, 1)]
    #[case(false, 0)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn register_check(#[case] check_enabled: bool, #[case] expected_count: usize) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = false;
        let context = AppContext::test(Some(config), None, None).unwrap();

        let mut check: MockHealthCheck = MockHealthCheck::default();
        check.expect_enabled().return_const(check_enabled);
        check.expect_name().return_const("test".to_string());

        // Act
        let mut subject: HealthCheckRegistry = HealthCheckRegistry::new(&context);
        subject.register(check).unwrap();

        // Assert
        assert_eq!(subject.checks().len(), expected_count);
        assert_eq!(
            subject.checks().iter().any(|check| check.name() == "test"),
            check_enabled
        );
    }
}
