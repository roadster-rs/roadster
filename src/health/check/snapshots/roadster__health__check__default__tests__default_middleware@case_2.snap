---
source: src/health/check/default.rs
expression: health_checks
---
[
    'db',
    'sidekiq-enqueue',
    'sidekiq-fetch',
    'smtp',
]
